design-proposals/klip-5-internal-schemas-for-KSQL-engine.md

```diff
@@ -0,0 +1,204 @@
+schema for the above schema will be as the following:
+
+```sql
+(COL_0 BIGINT, COL_1 STRING, COL_2 DOUBLE)
+```
+
+Note that the following one to one mapping will exist between the fields from the external schema and the fields from the internal schema:
+
+* “@ID”                            <==>   COL_0
+* “@NAME”                     <==>   COL_1
+* “MESSAGE.AMOUNT” <==>   COL_2
+
+In the case of nested fields, we can apply the same approach to the nested fields inside ARRAY, MAP
+and STRUCT types.
+
+Now that we have the internal schema built, the engine will rewrite the queries by replacing the
+identifiers that refer to the fields in the external schema with their corresponding field name in
+the internal schema.
+
+As an example, consider the following query again:
+
+```sql
+CREATE STREAM bar AS
+SELECT “@ID”, lowercase(foo.“@NAME”) AS “@NAME”, “MESSAGE.AMOUNT” * 100 AS “AMOUNT.BY.100”
+FROM foo
+WHERE “MESSAGE.AMOUNT” < 1;
+```
+
+With the above mapping the engine will write this query as the following:
+
+```sql
+
+CREATE STREAM bar AS
+SELECT FOO_COL_1 AS “@ID”, lowercase(FOO_COL_1) AS “@NAME”, FOO_COL_2 * 100 AS “AMOUNT.BY.100”
+FROM foo
+WHERE FOO_COL_2 < 1;
+```
+
+KSQL engine will be able to run the above query since all identifiers conform to the identifier
+naming requirements in KSQL. Note that KSQL already appends the source name to the field name
+automatically which will eliminate the ambiguity in case of having multiple sources (such as Join queries)
+ with the same field names.
+
+The sink schema for the above query will use the provided aliases for select expression as teh field names.
+So the sink schema will be the following:
+
+```sql
+(“@ID” BIGINT, “@NAME” STRING, “AMOUNT.BY.100” DOUBLE)
+```
+
+The engine will use the generated internal schema to build the streams application. The deserializers
+ from the source stream or table will use the external schema to fetch the fields and build the
+ GenericRow object, but beyond this point the internal schema will be used to process the records
+ for the queries.
+
+Also note that the result schema will be generated from the external schemas which means if we select
+any field from the external schema in the SELECT expressions and user does not specify an alias for
+the result column, the engine will use the same name from the external schema for the corresponding
+field in the result schema.
+
+The engine should verify the result schema field names comply with the requirements of the sink format.
+This verification should be done in query compile time. For instance, if the sink format is Avro, and the
+field names in the result schema do not comply with Avro naming requirments, KSQL engine should not run
+the query.
+
+Note that all the KSQL error or log messages should still use the external schema names since user
+does not know about the internal schemas.
+
+## Test Plan
+
+Tests for supporting quoted identifiers will be added.
```

Comment:
Other thoughts for the test plan:

1. Ensure we have test coverage for quoted identifiers that include periods in their name, e.g. `FOO.x`.  We currently _encode_ the source name into fields and this is likely to either not work, or be really ambiguous / error prone if we start to allow periods in identifiers.  Hence my recommendation that we implement our own Field type with explicit handling of fully qualified names.

2. Ensure we have test coverage of queries that select fields named ROWKEY and ROWTIME.  At the moment `KsqlStructuredDataSourceNode` will filter any such fields out.  We'll have to re-think this functionality if we're to support something like:

```sql
CREATE STREAM x AS SELECT "ROWTIME", y AS "ROWKEY" FROM Z;
```

We may choose to initially allow quoted identifiers to be reserved words, except `ROWTIME` and `ROWKEY`. But this will need documenting / testing.

Reply:
We should be able to use `"ROWKEY"` and `"ROWTIME"` since they are the same as any other reserved word and as long as they are quoted KSQL should be able to use treat them as any other quoted identifier.